compiler/vm/vmdef.nim

```diff
@@ -90,27 +90,138 @@ type
                       # minimize allocations; however the VM supports arbitrary
                       # temporary slot usage. This is required for the parameter
                       # passing implementation.
-    slotEmpty,        # slot is unused
-    slotFixedVar,     # slot is used for a fixed var/result (requires copy then)
-    slotFixedLet,     # slot is used for a fixed param/let
-    slotTempUnknown,  # slot but type unknown (argument of proc call)
-    slotTempInt,      # some temporary int
-    slotTempFloat,    # some temporary float
-    slotTempStr,      # some temporary string
-    slotTempComplex,  # some complex temporary (s.node field is used)
-    slotTempPerm      # slot is temporary but permanent (hack)
-
-
-  TFullReg* = object  # with a custom mark proc, we could use the same
-                      # data representation as LuaJit (tagged NaNs).
-    case kind*: TRegisterKind
-    of rkNone: nil
-    of rkInt: intVal*: BiggestInt
-    of rkFloat: floatVal*: BiggestFloat
-    of rkNode: node*: PNode
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    slotEmpty,        ## slot is unused
+    slotFixedVar,     ## slot is used for a fixed var/result (requires copy then)
+    slotFixedLet,     ## slot is used for a fixed param/let
+    slotTempUnknown,  ## slot but type unknown (argument of proc call)
+    slotTempInt,      ## some temporary int
+    slotTempFloat,    ## some temporary float
+    slotTempStr,      ## some temporary string
+    slotTempComplex,  ## some complex temporary (s.node field is used)
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    akInt
+    akFloat
+    akString # TODO: maybe even collapse akString into akOrdinalArray?
+    akOrdinalArray # Space efficient storing of float/int arrays
+    akArray # Used for both seqs and fixed size arrays by vmgen
+    akSet
+    akRef
+    akPtr
+    akPointer ## A raw pointer points to inside akString's string or akOrdinalArray/akArray's rawArrayVal
+    akObject
+    akPNode
+    akCallable
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
+    of ckTemplate:
+      templateSym*: PSym
+
+
+  OrdinalArrayKind* = enum
+    oakUnsigned
+    oakSigned
+    oakFloat32
+    oakFloat
+
+  VmObjectAtom* = object
+    case kind*: AtomKind
+    of akInt:
+      # XXX: is it okay to use this for unsigned integer too?
+      intVal*: BiggestInt
```

Comment:
It's probably better to just use unsigned here and do the `int` checks when the instruction calls for it.